File: auth0_flutter/darwin/Classes/MyAccountAPI/MyAccountExtensions.swift
import Auth0

#if os(iOS)
import Flutter
#else
import FlutterMacOS
#endif

extension FlutterError {
    convenience init(from error: MyAccountError) {
        let details: [String: Any] = [
            "_statusCode": error.statusCode,
            "_title": error.title,
            "_detail": error.detail,
            "_errorFlags": [
                "isNetworkError": error.isNetworkError
            ]
        ]
        self.init(code: error.code,
                  message: error.detail.isEmpty ? error.title : error.detail,
                  details: details)
    }
}

extension AuthenticationMethod {
    func asDictionary() -> [String: Any?] {
        return [
            "id": id,
            "type": type,
            "name": name,
            "phone_number": phoneNumber,
            "email": nil,
            "totp_secret": nil,
            "totp_uri": nil,
            "preferred_authentication_method": preferredAuthenticationMethod,
            "created_at": createdAt,
            "last_auth_at": nil,
            "confirmed": confirmed,
            "usage": usage
        ]
    }
}

extension PhoneEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "id": authenticationId,
            "auth_session": authenticationSession,
            "totp_secret": nil,
            "totp_uri": nil,
            "barcode_uri": nil,
            "recovery_code": nil
        ]
    }
}

extension EmailEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "id": authenticationId,
            "auth_session": authenticationSession,
            "totp_secret": nil,
            "totp_uri": nil,
            "barcode_uri": nil,
            "recovery_code": nil
        ]
    }
}

extension TOTPEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "id": authenticationId,
            "auth_session": authenticationSession,
            "totp_secret": authenticatorManualInputCode,
            "totp_uri": authenticatorQRCodeURI,
            "barcode_uri": authenticatorQRCodeURI,
            "recovery_code": nil
        ]
    }
}

extension PushEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "id": authenticationId,
            "auth_session": authenticationSession,
            "totp_secret": nil,
            "totp_uri": nil,
            "barcode_uri": authenticatorQRCodeURI,
            "recovery_code": nil
        ]
    }
}

extension RecoveryCodeEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        return [
            "id": authenticationId,
            "auth_session": authenticationSession,
            "totp_secret": nil,
            "totp_uri": nil,
            "barcode_uri": nil,
            "recovery_code": recoveryCode
        ]
    }
}

#if PASSKEYS_PLATFORM
@available(iOS 16.6, macOS 13.5, visionOS 1.0, *)
extension PasskeyEnrollmentChallenge {
    func asDictionary() -> [String: Any?] {
        // Forward the WebAuthn creation options in the flat shape the app's
        // platform-authenticator ceremony consumes, plus the authentication
        // method id needed to finish enrollment.
        return [
            "authenticationMethodId": authenticationMethodId,
            "authSession": authenticationSession,
            "authParamsPublicKey": [
                "challenge": challengeData.base64URLEncodedString(),
                "rpId": relyingPartyId,
                "userId": userId.base64URLEncodedString(),
                "userName": userName
            ]
        ]
    }
}

@available(iOS 16.6, macOS 13.5, visionOS 1.0, *)
extension PasskeyAuthenticationMethod {
    func asDictionary() -> [String: Any?] {
        let createdAtFormatter = ISO8601DateFormatter()
        createdAtFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        return [
            "id": id,
            "type": type,
            "identity_user_id": userIdentityId,
            "user_agent": userAgent,
            "key_id": credential.id,
            "public_key": credential.publicKey.base64EncodedString(),
            "user_handle": credential.userHandle.base64URLEncodedString(),
            "credential_device_type": credential.deviceType.rawValue,
            "credential_backed_up": credential.isBackedUp,
            "aaguid": aaguid,
            "relying_party_id": relyingPartyIdentifier,
            "created_at": createdAtFormatter.string(from: createdAt)
        ]
    }
}
#endif

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/MyAccountExtensions.kt
package com.auth0.auth0_flutter

import com.auth0.android.result.AuthenticationMethod
import com.auth0.android.result.EmailAuthenticationMethod
import com.auth0.android.result.EnrollmentChallenge
import com.auth0.android.result.MfaAuthenticationMethod
import com.auth0.android.result.MfaEnrollmentChallenge
import com.auth0.android.result.PasskeyAuthenticationMethod
import com.auth0.android.result.PasskeyEnrollmentChallenge
import com.auth0.android.result.PhoneAuthenticationMethod
import com.auth0.android.result.PushNotificationAuthenticationMethod
import com.auth0.android.result.RecoveryCodeEnrollmentChallenge
import com.auth0.android.result.TotpAuthenticationMethod
import com.auth0.android.result.TotpEnrollmentChallenge
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

fun AuthenticationMethod.toMyAccountMethodMap(): Map<String, Any?> {
    return buildMap {
        put("id", id)
        put("type", type)
        put("created_at", createdAt)
        when (val method = this@toMyAccountMethodMap) {
            is PhoneAuthenticationMethod -> {
                put("name", method.name)
                put("phone_number", method.phoneNumber)
                put("preferred_authentication_method",
                    method.preferredAuthenticationMethod)
            }
            is EmailAuthenticationMethod -> {
                put("name", method.name)
                put("email", method.email)
            }
            is TotpAuthenticationMethod -> {
                put("name", method.name)
            }
            is PushNotificationAuthenticationMethod -> {
                put("name", method.name)
            }
            else -> {
                put("name", null)
            }
        }
        put("usage", usage)
        if (this@toMyAccountMethodMap is MfaAuthenticationMethod) {
            put("confirmed", confirmed)
        }
    }
}

fun PasskeyAuthenticationMethod.toMyAccountPasskeyMethodMap(): Map<String, Any?> {
    return buildMap {
        put("id", id)
        put("type", type)
        put("created_at", createdAt)
        put("usage", usage)
        put("identity_user_id", identityUserId)
        put("user_agent", userAgent)
        put("key_id", keyId)
        put("public_key", publicKey)
        put("user_handle", userHandle)
        put("credential_device_type", credentialDeviceType)
        put("credential_backed_up", credentialBackedUp)
        put("transports", transports)
        put("aaguid", aaguid)
        put("relying_party_id", relyingPartyId)
    }
}

fun PasskeyEnrollmentChallenge.toMyAccountPasskeyChallengeMap(): Map<String, Any?> {
    val gson = Gson()
    // AuthnParamsPublicKey is a typed data class; convert via JSON tree to avoid
    // a toJson()/fromJson(String) round-trip through an intermediate String.
    val authParamsPublicKeyMap: Map<String, Any> = gson.fromJson(
        gson.toJsonTree(authParamsPublicKey),
        object : TypeToken<Map<String, Any>>() {}.type
    )
    return buildMap {
        put("authenticationMethodId", authenticationMethodId)
        put("authSession", authSession)
        put("authParamsPublicKey", authParamsPublicKeyMap)
    }
}

fun EnrollmentChallenge.toMyAccountChallengeMap(): Map<String, Any?> {
    return buildMap {
        put("id", id)
        put("auth_session", authSession)
        when (val challenge = this@toMyAccountChallengeMap) {
            is TotpEnrollmentChallenge -> {
                put("barcode_uri", challenge.barcodeUri)
                put("totp_secret", challenge.manualInputCode)
                put("totp_uri", challenge.barcodeUri)
            }
            is RecoveryCodeEnrollmentChallenge -> {
                put("recovery_code", challenge.recoveryCode)
            }
            is MfaEnrollmentChallenge -> {}
            else -> {}
        }
    }
}
